fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/MetaDataProtoEditor.java

```diff
@@ -174,16 +363,19 @@ public static void addNestedRecordType(@Nonnull RecordMetaDataProto.MetaData.Bui
      * @param metaDataBuilder the meta-data builder
      * @param recordType the record type to be deprecated
      */
-    @Nonnull
     public static void deprecateRecordType(@Nonnull RecordMetaDataProto.MetaData.Builder metaDataBuilder, @Nonnull String recordType) {
-        DescriptorProtos.DescriptorProto.Builder unionBuilder = fetchUnionBuilder(metaDataBuilder.getRecordsBuilder());
+        final DescriptorProtos.FileDescriptorProto.Builder fileBuilder = metaDataBuilder.getRecordsBuilder();
+        DescriptorProtos.DescriptorProto.Builder unionBuilder = fetchUnionBuilder(fileBuilder);
         if (unionBuilder.getName().equals(recordType)) {
             throw new MetaDataException("Cannot deprecate the union");
         }
         // deprecate all fields of type recordType from the union.
         boolean found = false;
         for (DescriptorProtos.FieldDescriptorProto.Builder fieldBuilder : unionBuilder.getFieldBuilderList()) {
-            if (fieldBuilder.getTypeName().equals(recordType)) {
+            final FieldTypeMatch fieldTypeMatch = fieldIsType(fileBuilder, unionBuilder, fieldBuilder, recordType);
+            if (fieldTypeMatch.isPossibleMatch()) {
+                throw new AmbiguousTypeNameException(fileBuilder.getPackage(), unionBuilder, fieldBuilder, fullyQualifiedTypeName(fileBuilder, recordType));
+            } else if (FieldTypeMatch.MATCHES.equals(fieldTypeMatch)) {
```

Comment:
Could `MATCHES_AS_NESTED` happen? Maybe somehow, either thru the comments or code, make this case explicit?

Reply:
> It's somewhat unclear what that would mean. It would require that the user do something like add a field to the union which is defined as a nested type within a different record...which is somewhat weird, but could happen.

What you have is ok. We could disallow it too.

Reply:
I think I prefer deprecating the type like this to disallowing types that are defined as nested types as top-level record types. This is partially because detecting this is somewhat difficult, and though it's weird, I'm not sure it costs us anything to support it. But maybe @MMcM has opinions on that?

Reply:
Sure, as long as the contract is explicit I'm fine with it. Just to clarify - I meant disallowing `add a field to the union which is defined as a nested type within a different record` not `nested types as top-level record types`.